### PR TITLE
New PlayProjectCard deferred action + Prelude cards refactor

### PR DESCRIPTION
**Pull request by Lynesth**
Changes:

**PlayProjectCard deferred action <sup>NEW</sup>**
Used when you play a card or use an action that makes you play a card from your hand (e.g. Eccentric Sponsor).

**DrawCards deferred action**
Can now be used to draw specific tag cards. Also logs simple card draw now `PLAYER1 drew X card(s)`.

**PlaceCityTile deferred action**
Can now receive a list of available spaces for special cities (like Lava Tube Settlement).

Those changes have been applied to cards from the Prelude extension for now. More to come in futur PRs.

### CODE REVIEW

File: src/cards/prelude/LavaTubeSettlement.ts [hunk 1]
```diff
@@ -25,19 +24,20 @@ export class LavaTubeSettlement implements IProjectCard {
...
-            (space: ISpace) => {
-                game.addCityTile(player, space.id);
-                player.addProduction(Resources.MEGACREDITS,2);
-                player.addProduction(Resources.ENERGY,-1);
-                return undefined;
-            }
-        );
+        player.addProduction(Resources.MEGACREDITS,2);
+        player.addProduction(Resources.ENERGY,-1);
+        game.defer(new PlaceCityTile(
```
Comment (kberg) [inline src/cards/prelude/LavaTubeSettlement.ts:35]: You aren't responsible for this, but - huh - where does limiting the tiles to the volcanic sites happen?

File: src/deferredActions/DrawCards.ts [hunk 2]
```diff
@@ -1,18 +1,33 @@
...
+        public tag?: Tags
     ){}
 
     public execute() {
-        for (let i = 0; i < this.count; i++) {
-            this.player.cardsInHand.push(this.game.dealer.dealCard());
+        const drawnCards: Array<IProjectCard> = [];
+
+        if (this.tag !== undefined) {
+            for (const foundCard of this.game.drawCardsByTag(this.tag, 2)) {
```
Comment (kberg) [inline src/deferredActions/DrawCards.ts:19]: This top half doesn't make much sense.

File: tests/cards/prelude/AcquiredSpaceAgency.spec.ts [hunk 3]
```diff
@@ -11,8 +10,12 @@ describe("AcquiredSpaceAgency", function () {
         const card = new AcquiredSpaceAgency();
         const player = new Player("test", Color.BLUE, false);
         const game = new Game("foobar", [player], player);
-        const action = card.play(player, game);
-        expect(action).to.eq(undefined);
+        card.play(player, game);
+        expect(game.deferredActions.length).to.eq(1);
+
+        // Draw cards
+        game.runDeferredAction(game.deferredActions[0], () => {});
```
Comment (kberg) [inline tests/cards/prelude/AcquiredSpaceAgency.spec.ts:17]: This is absolutely worth a method on its own.

File: tests/cards/prelude/EarlySettlement.spec.ts [hunk 4]
```diff
@@ -1,24 +1,26 @@
...
         const card = new EarlySettlement();
         const player = new Player("test", Color.BLUE, false);
         const game = new Game("foobar", [player], player);
-        const action = card.play(player, game) as SelectSpace;
-        expect(action).not.to.eq(undefined);
+
+        card.play(player, game);
+        const selectSpace = game.deferredActions[0].execute() as SelectSpace;
+
+        expect(selectSpace).not.to.eq(undefined);
```
Comment (kberg) [inline tests/cards/prelude/EarlySettlement.spec.ts:19]: .is.not.undefined.

**Review by kberg — CHANGES_REQUESTED** [review]
Minor comments.

File: src/cards/prelude/LavaTubeSettlement.ts (hunk 1 above)
Comment (Lynesth) [reply src/cards/prelude/LavaTubeSettlement.ts:35]: In the `getSpacesForCity` method of the card.

File: tests/cards/prelude/AcquiredSpaceAgency.spec.ts (hunk 3 above)
Comment (Lynesth) [reply tests/cards/prelude/AcquiredSpaceAgency.spec.ts:17]: There quite a few of those in the tests.
I can make a PR later that just changes all of them if you want.

File: src/deferredActions/DrawCards.ts (hunk 2 above)
Comment (Lynesth) [reply src/deferredActions/DrawCards.ts:19]: This is totally my bad, I messed up when splitting the PR in two. Will fix.

File: tests/cards/prelude/EarlySettlement.spec.ts (hunk 4 above)
Comment (kberg) [reply tests/cards/prelude/EarlySettlement.spec.ts:19]: Nah, I wrote #1766

File: src/cards/prelude/LavaTubeSettlement.ts (hunk 1 above)
Comment (kberg) [reply src/cards/prelude/LavaTubeSettlement.ts:35]: Gotcha, thanks.

File: tests/cards/prelude/AcquiredSpaceAgency.spec.ts (hunk 3 above)
Comment (kberg) [reply tests/cards/prelude/AcquiredSpaceAgency.spec.ts:17]: Later is good.

File: tests/cards/prelude/EarlySettlement.spec.ts (hunk 4 above)
Comment (Lynesth) [reply tests/cards/prelude/EarlySettlement.spec.ts:19]: There's a lot of those in the tests.
I can make a PR later that just changes all of them if you want and if that's the correct way of doing it (I don't no chaijs much).
I've noticed some `is.length()` for example which shouldn't be used anymore as it's deprecated.

File: src/deferredActions/DrawCards.ts (hunk 2 above)
Comment (Lynesth) [reply src/deferredActions/DrawCards.ts]: I fixed it. Let me know if it makes sens to you know.

File: src/deferredActions/DrawCards.ts [hunk 5]
```diff
@@ -16,17 +16,30 @@ export class DrawCards implements DeferredAction {
...
         if (this.tag !== undefined) {
-            for (const foundCard of this.game.drawCardsByTag(this.tag, 2)) {
+            // Reveal cards from the deck until |this.count| of a specific tag have be drawn
+            for (const foundCard of this.game.drawCardsByTag(this.tag, this.count)) {
                 drawnCards.push(foundCard);
             }
-            this.game.log("${0} drew ${1} and ${2}", b => b.player(this.player).card(drawnCards[0]).card(drawnCards[1]));
+            switch (this.count) {
+                case 1:
+                    this.game.log("${0} drew ${1} and ${2}", b => b.player(this.player).card(drawnCards[0]));
```
Comment (nwai90) [inline src/deferredActions/DrawCards.ts:25]: This should be `this.game.log("${0} drew ${1}", b => b.player(this.player).card(drawnCards[0]));` as there is only one card

Comment (Lynesth) [reply src/deferredActions/DrawCards.ts]: I have no idea what you are talking about.... ;)

File: src/deferredActions/DrawCards.ts [hunk 6]
```diff
@@ -1,18 +1,46 @@
...
-            this.player.cardsInHand.push(this.game.dealer.dealCard());
+        const drawnCards: Array<IProjectCard> = [];
+
+        if (this.tag !== undefined) {
+            // Reveal cards from the deck until |this.count| of a specific tag have be drawn
+            for (const foundCard of this.game.drawCardsByTag(this.tag, this.count)) {
+                drawnCards.push(foundCard);
+            }
+            switch (this.count) {
+                case 1:
```
Comment (kberg) [inline src/deferredActions/DrawCards.ts:24]: Can be reduced to 

```
// if count = 3, for instance, this generates "${0} drew ${1}, ${2}, ${3}"
const msg = "${0} drew " + ([...Array(this.count).keys()].map(n => "${" + (n+1) + "}").join(", "));
this.game.log(msg, b => drawnCards.forEach(c => b.card(c)));
```

**Review by kberg — APPROVED** [review]
Your call if you want to make that change. But at least do something sensible if count > 3, please.

File: src/deferredActions/DrawCards.ts (hunk 6 above)
Comment (Lynesth) [reply src/deferredActions/DrawCards.ts:24]: There's no card making you draw more than 3 specific tag cards that why I stopped at 3.
But I'm more than happy to go with your idea.

File: src/deferredActions/DrawCards.ts (hunk 2 above)
Comment (kberg) [reply src/deferredActions/DrawCards.ts]: Thanks!

**Review by kberg — APPROVED** [review]

**Pull request merged**